chainstate/src/detail/mod.rs

```diff
@@ -271,52 +274,179 @@ impl<S: BlockchainStorage, V: TransactionVerificationStrategy> Chainstate<S, V>
         &mut self,
-        block: WithId<Block>,
-        block_source: BlockSource,
-    ) -> Result<Option<BlockIndex>, BlockError> {
-        let block = self.check_legitimate_orphan(block_source, block).log_err()?;
-
-        let block_id = block.get_id();
+        mut func: Func,
+        block_id: &Id<Block>,
+        comment: &str,
+    ) -> Result<Res, BlockError>
+    where
+        Func: FnMut(&mut chainstateref::ChainstateRef<TxRw<'_, S>, V>) -> Result<Res, BlockError>,
+    {
         let mut attempt_number = 0;
         loop {
-            log::info!("Processing block: {}", block_id);
+            log::info!("Processing block ({}), block id = {}", comment, block_id);
 
             let mut chainstate_ref = self.make_db_tx().map_err(BlockError::from).log_err()?;
 
-            let best_block_id = chainstate_ref
-                .get_best_block_id()
-                .map_err(BlockError::BestBlockLoadError)
-                .log_err()?;
-
-            chainstate_ref
-                .check_block(&block, false)
-                .map_err(BlockError::CheckBlockFailed)
-                .log_err()?;
+            let result = func(&mut chainstate_ref).log_err()?;
 
-            let block_index = chainstate_ref.persist_block(&block).log_err()?;
-            let result =
-                chainstate_ref.activate_best_chain(block_index, best_block_id).log_err()?;
             let db_commit_result = chainstate_ref.commit_db_tx().log_err();
             match db_commit_result {
-                Ok(_) => {}
+                Ok(_) => return Ok(result),
                 Err(err) => {
                     if attempt_number >= *self.chainstate_config.max_db_commit_attempts {
                         return Err(BlockError::DatabaseCommitError(
-                            block.get_id(),
+                            *block_id,
                             *self.chainstate_config.max_db_commit_attempts,
                             err,
+                            comment.into(),
                         ));
                     } else {
                         attempt_number += 1;
                         continue;
                     }
                 }
             }
+        }
+    }
+
+    /// This is similar to `with_rw_tx_for_block_id`, but it also maintains a mutable state that
+    /// is passed to `func` and then returned to the caller.
+    /// Note that the state is reset to `initial_state` on each commit attempt and the returned
+    /// state is the one from the last commit attempt.
+    fn with_rw_tx_and_state_for_block_id<Func, Res, State>(
+        &mut self,
+        mut func: Func,
+        initial_state: &State,
+        block_id: &Id<Block>,
+        comment: &str,
+    ) -> (State, Result<Res, BlockError>)
+    where
+        Func: FnMut(
+            &mut chainstateref::ChainstateRef<TxRw<'_, S>, V>,
+            &mut State,
+        ) -> Result<Res, BlockError>,
+        State: Clone,
```

Comment:
> I am wondering if there is much value in a function like this. Since func is a FnMut, you can just mutably capture any state that needs to be modified from the outer scope (as the implementation here does). No extra argument is necessary.

Well, yeah, with the missing clone on each iteration is doesn't make much sense, but that was a bug, as I've already mentioned.

> If you want to make the state manipulation more functional in style, I'd avoid using &mut references to it. Just pass it in/out by value everywhere.

In general, the &mut parameter is needed by `perform_block_checks`, where I want to update the status without the need to return it explicitly. IMO without it, the logic of `perform_block_checks` would be more complicated.

Reply:
> > If you want to make the state manipulation more functional in style, I'd avoid using &mut references to it. Just pass it in/out by value everywhere.
> 
> In general, the &mut parameter is needed by `perform_block_checks`, where I want to update the status without the need to return it explicitly. IMO without it, the logic of `perform_block_checks` would be more complicated.

Fair point. Though my original suggestion was supposed to only apply to the generic helper function(s). Ultimately it's a matter of taste to a degree so I don't mind either way. The whole discussion was prompted by the initial confusion around `Clone` anyway.